### ISSUE TITLE
Is normalization of the images must?

### ISSUE THREAD

**Issue by saigopalm** [ultralytics/ultralytics#12726]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello! I am training a YOLOv8n-seg model on a custom dataset and the outputs I am getting are fully black. I performed resizing and data augmentation also. Is this because of not normalizing the input images? Is normalizing the images an implicit step expected to be done on images before training YOLOv8 model?

**Comment by glenn-jocher** [comment]
Hello! Yes, normalizing the input images is important when training YOLOv8 models. YOLOv8 expects image pixel values to be normalized. Typically, images should be normalized to have values between 0 and 1. This can be done by dividing the RGB values by 255. Here's a quick example of how you might normalize your images:

```python
# Example of normalizing an image
image = image / 255.0
```

Make sure to apply this normalization consistently across all images in your dataset during both training and inference stages. This could indeed resolve the issue with the outputs you are encountering. Give it a try and let us know if it helps! 😊

**Comment by saigopalm** [comment]
In opencv after normalization I am not able to save the images using cv2.imwrite. How do I save the images after normalization?

**Comment by glenn-jocher** [comment]
Hello!

After normalizing your images (values between 0 and 1), you'll need to convert them back to the 0-255 range before saving with `cv2.imwrite`. You can do this by multiplying the normalized image by 255 and converting it to an integer data type. Here's how you can do it:

```python
# Assume 'normalized_image' is your normalized image with values between 0 and 1
image_to_save = (normalized_image * 255).astype('uint8')
cv2.imwrite('path_to_save.jpg', image_to_save)
```

This will convert your image back to the appropriate format for saving. Let me know if this works for you! 😊

**Comment by ge1mina023** [comment]
> Hello! Yes, normalizing the input images is important when training YOLOv8 models. YOLOv8 expects image pixel values to be normalized. Typically, images should be normalized to have values between 0 and 1. This can be done by dividing the RGB values by 255. Here's a quick example of how you might normalize your images:
> 
> ```python
> # Example of normalizing an image
> image = image / 255.0
> ```
> 
> Make sure to apply this normalization consistently across all images in your dataset during both training and inference stages. This could indeed resolve the issue with the outputs you are encountering. Give it a try and let us know if it helps! 😊

We need to convert the image format from BGR to RGB, so is *. jpg also in RGB format during training?
@glenn-jocher

**Comment by glenn-jocher** [comment]
@ge1mina023 hello!

Great question! When working with images in OpenCV, they are typically loaded in BGR format. However, YOLOv8 models expect images in RGB format. If you are using OpenCV to load your images, you will need to convert them from BGR to RGB before normalizing and feeding them into the model.

Here's how you can do this conversion:

```python
import cv2

# Load the image in BGR format
image_bgr = cv2.imread('path_to_image.jpg')

# Convert the image from BGR to RGB
image_rgb = cv2.cvtColor(image_bgr, cv2.COLOR_BGR2RGB)

# Normalize the image
normalized_image = image_rgb / 255.0
```

Regarding your question about `.jpg` format during training, the format of the image file itself (e.g., `.jpg`, `.png`) does not dictate the color space. It's the library used to load the image that determines the color space. OpenCV loads images in BGR by default, while other libraries like PIL load images in RGB.

Ultralytics' YOLO models do not automatically convert images from BGR to RGB. Therefore, it's essential to handle this conversion in your preprocessing pipeline before passing the images to the model.

If you have any further questions or run into any issues, feel free to ask! 😊

**Comment by ge1mina023** [comment]
> @ge1mina023 hello!
> 
> Great question! When working with images in OpenCV, they are typically loaded in BGR format. However, YOLOv8 models expect images in RGB format. If you are using OpenCV to load your images, you will need to convert them from BGR to RGB before normalizing and feeding them into the model.
> 
> Here's how you can do this conversion:
> 
> ```python
> import cv2
> 
> # Load the image in BGR format
> image_bgr = cv2.imread('path_to_image.jpg')
> 
> # Convert the image from BGR to RGB
> image_rgb = cv2.cvtColor(image_bgr, cv2.COLOR_BGR2RGB)
> 
> # Normalize the image
> normalized_image = image_rgb / 255.0
> ```
> 
> Regarding your question about `.jpg` format during training, the format of the image file itself (e.g., `.jpg`, `.png`) does not dictate the color space. It's the library used to load the image that determines the color space. OpenCV loads images in BGR by default, while other libraries like PIL load images in RGB.
> 
> Ultralytics' YOLO models do not automatically convert images from BGR to RGB. Therefore, it's essential to handle this conversion in your preprocessing pipeline before passing the images to the model.
> 
> If you have any further questions or run into any issues, feel free to ask! 😊

Does it mean that any *. jpg image saved on the hard drive does not have a specific color channel, and this is just a behavior of some libraries loading and saving in memory? So, what is the channel order in which the images used for training in the disk are loaded into the GPU for training? @glenn-jocher

**Comment by glenn-jocher** [comment]
Hello @ge1mina023,

Great follow-up question! You are correct that the `.jpg` format itself does not inherently specify a color channel order. The color channel order is determined by the library used to load the image into memory.

When using OpenCV, images are loaded in BGR format by default. However, other libraries like PIL load images in RGB format. This is why it's crucial to ensure that your images are in the correct color space before feeding them into the YOLOv8 model.

For training, the images should be in RGB format. Here’s a quick summary of the process:

1. **Load the image using OpenCV (BGR format):**
   ```python
   image_bgr = cv2.imread('path_to_image.jpg')
   ```

2. **Convert the image from BGR to RGB:**
   ```python
   image_rgb = cv2.cvtColor(image_bgr, cv2.COLOR_BGR2RGB)
   ```

3. **Normalize the image:**
   ```python
   normalized_image = image_rgb / 255.0
   ```

When these images are loaded into the GPU for training, they should be in the RGB format. This ensures that the model interprets the color channels correctly.

If you encounter any further issues or have more questions, feel free to ask! 😊

**Comment by ge1mina023** [comment]
> Hello @ge1mina023,
> 
> Great follow-up question! You are correct that the `.jpg` format itself does not inherently specify a color channel order. The color channel order is determined by the library used to load the image into memory.
> 
> When using OpenCV, images are loaded in BGR format by default. However, other libraries like PIL load images in RGB format. This is why it's crucial to ensure that your images are in the correct color space before feeding them into the YOLOv8 model.
> 
> For training, the images should be in RGB format. Here’s a quick summary of the process:
> 
> 1. **Load the image using OpenCV (BGR format):**
>    ```python
>    image_bgr = cv2.imread('path_to_image.jpg')
>    ```
> 2. **Convert the image from BGR to RGB:**
>    ```python
>    image_rgb = cv2.cvtColor(image_bgr, cv2.COLOR_BGR2RGB)
>    ```
> 3. **Normalize the image:**
>    ```python
>    normalized_image = image_rgb / 255.0
>    ```
> 
> When these images are loaded into the GPU for training, they should be in the RGB format. This ensures that the model interprets the color channels correctly.
> 
> If you encounter any further issues or have more questions, feel free to ask! 😊

I don't understand at all. The method you mentioned was already in use when I inference the images, but it seems that there was no training available. It was just using simple JPG images for Ultratics training, and I didn't have any chance to do any preprocessing. I seem to have found a parameter bgr in the [document,](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters) how do I use it? @glenn-jocher

**Comment by glenn-jocher** [comment]
Hello @ge1mina023,

Thank you for your patience and for pointing out the `bgr` parameter in the documentation. You are correct; the `bgr` parameter is designed to handle images loaded in BGR format directly within the training pipeline.

To use the `bgr` parameter, you can set it in your training configuration. This will ensure that the images are correctly interpreted as BGR without needing manual conversion. Here’s how you can do it:

1. **Set the `bgr` parameter in your training command:**
   ```bash
   yolo train data=coco128.yaml model=yolov8n.pt epochs=100 imgsz=640 bgr=True
   ```

2. **Or, set it in your YAML configuration file:**
   ```yaml
   # Example configuration in your data.yaml or custom config file
   bgr: True
   ```

This should handle the color space conversion automatically, allowing you to use your images as they are without additional preprocessing.

If you continue to experience issues, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example would be very helpful for us to investigate further. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions! 😊

**Comment by ge1mina023** [comment]
> Hello @ge1mina023,
> 
> Thank you for your patience and for pointing out the `bgr` parameter in the documentation. You are correct; the `bgr` parameter is designed to handle images loaded in BGR format directly within the training pipeline.
> 
> To use the `bgr` parameter, you can set it in your training configuration. This will ensure that the images are correctly interpreted as BGR without needing manual conversion. Here’s how you can do it:
> 
> 1. **Set the `bgr` parameter in your training command:**
>    ```shell
>    yolo train data=coco128.yaml model=yolov8n.pt epochs=100 imgsz=640 bgr=True
>    ```
> 2. **Or, set it in your YAML configuration file:**
>    ```yaml
>    # Example configuration in your data.yaml or custom config file
>    bgr: True
>    ```
> 
> This should handle the color space conversion automatically, allowing you to use your images as they are without additional preprocessing.
> 
> If you continue to experience issues, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example would be very helpful for us to investigate further. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 
> Feel free to reach out if you have any more questions! 😊

So, if we don't use the bgr parameter, what is the channel order of loading *. jpg images for training in Ultratics? Because you mentioned that the *. jpg images on the disk do not have a specific channel order. @glenn-jocher

**Comment by glenn-jocher** [comment]
Hello @ge1mina023,

Great question! If you don't use the `bgr` parameter, Ultralytics YOLO models expect images to be in RGB format by default. This means that when you load images for training, they should be in RGB order. If you are using OpenCV to load images (which loads them in BGR format), you would need to convert them to RGB before training unless you set `bgr=True`.

To summarize:
- **Without `bgr=True`**: Ensure images are in RGB format before training.
- **With `bgr=True`**: You can load images in BGR format, and the model will handle the conversion.

If you have any further questions or need additional assistance, feel free to ask! 😊
